# Name: Yuning Lei
# Github username: Thundereggg
# USC ID: 4169712364

In [1]:
import pandas as pd  
import numpy as np   
from sklearn.model_selection import train_test_split
from sklearn.metrics import hamming_loss
import seaborn as sns 
import matplotlib.pyplot as plt  
from sklearn.model_selection import GridSearchCV
from sklearn import svm
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC
from sklearn import preprocessing
from sklearn.svm import LinearSVC
from imblearn.over_sampling import SMOTE
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from scipy.cluster.hierarchy import dendrogram, linkage
from scipy.spatial.distance import squareform

## 1. Multi-class and Multi-Label Classification Using Support Vector Machines

(a) Download the Anuran Calls (MFCCs) Data Set from: https://archive.ics. uci.edu/ml/datasets/Anuran+Calls+%28MFCCs%29. Choose 70% of the data randomly as the training set.

In [2]:
data=pd.read_csv('../data/AnuranCalls(MFCCs)/Frogs_MFCCs.csv')
data

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
0          1.0  0.152936 -0.105586  0.200722  0.317201  0.260764  0.100945   
1          1.0  0.171534 -0.098975  0.268425  0.338672  0.268353  0.060835   
2          1.0  0.152317 -0.082973  0.287128  0.276014  0.189867  0.008714   
3          1.0  0.224392  0.118985  0.329432  0.372088  0.361005  0.015501   
4          1.0  0.087817 -0.068345  0.306967  0.330923  0.249144  0.006884   
...        ...       ...       ...       ...       ...       ...       ...   
7190       1.0 -0.554504 -0.337717  0.035533  0.034511  0.443451  0.093889   
7191       1.0 -0.517273 -0.370574  0.030673  0.068097  0.402890  0.096628   
7192       1.0 -0.582557 -0.343237  0.029468  0.064179  0.385596  0.114905   
7193       1.0 -0.519497 -0.307553 -0.004922  0.072865  0.377131  0.086866   
7194       1.0 -0.508833 -0.324106  0.062068  0.078211  0.397188  0.094596   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_17  MFCCs_18  MFCCs_19  \
0    -0.150063 -0.171128  0.124676  ... -0.108351 -0.077623 -0.009568   
1    -0.222475 -0.207693  0.170883  ... -0.090974 -0.056510 -0.035303   
2    -0.242234 -0.219153  0.232538  ... -0.050691 -0.023590 -0.066722   
3    -0.194347 -0.098181  0.270375  ... -0.136009 -0.177037 -0.130498   
4    -0.265423 -0.172700  0.266434  ... -0.048885 -0.053074 -0.088550   
...        ...       ...       ...  ...       ...       ...       ...   
7190 -0.100753  0.037087  0.081075  ...  0.069430  0.071001  0.021591   
7191 -0.116460  0.063727  0.089034  ...  0.061127  0.068978  0.017745   
7192 -0.103317  0.070370  0.081317  ...  0.082474  0.077771 -0.009688   
7193 -0.115799  0.056979  0.089316  ...  0.051796  0.069073  0.017963   
7194 -0.117672  0.058874  0.076180  ...  0.061455  0.072983 -0.003980   

      MFCCs_20  MFCCs_21  MFCCs_22           Family      Genus  \
0     0.057684  0.118680  0.014038  Leptodactylidae  Adenomera   
1     0.020140  0.082263  0.029056  Leptodactylidae  Adenomera   
2    -0.025083  0.099108  0.077162  Leptodactylidae  Adenomera   
3    -0.054766 -0.018691  0.023954  Leptodactylidae  Adenomera   
4    -0.031346  0.108610  0.079244  Leptodactylidae  Adenomera   
...        ...       ...       ...              ...        ...   
7190  0.052449 -0.021860 -0.079860          Hylidae     Scinax   
7191  0.046461 -0.015418 -0.101892          Hylidae     Scinax   
7192  0.027834 -0.000531 -0.080425          Hylidae     Scinax   
7193  0.041803 -0.027911 -0.096895          Hylidae     Scinax   
7194  0.031560 -0.029355 -0.087910          Hylidae     Scinax   

             Species  RecordID  
0     AdenomeraAndre         1  
1     AdenomeraAndre         1  
2     AdenomeraAndre         1  
3     AdenomeraAndre         1  
4     AdenomeraAndre         1  
...              ...       ...  
7190     ScinaxRuber        60  
7191     ScinaxRuber        60  
7192     ScinaxRuber        60  
7193     ScinaxRuber        60  
7194     ScinaxRuber        60  

[7195 rows x 26 columns]

In [3]:
x=data.drop(['Family','Genus','Species','RecordID'],axis=1)
y=data[['Family','Genus','Species']]
train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3,train_size=0.7, random_state=42)
train_x

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
2004  1.000000  0.327476  0.282440  0.572211  0.096509 -0.036705 -0.082174   
1194  1.000000  0.232818 -0.144331 -0.078909 -0.072467  0.255380  0.450215   
5359  0.966443  0.455914  1.000000  0.541650 -0.412873  0.307099  0.235819   
1756  1.000000  0.257873  0.112807  0.527614  0.186094  0.043683 -0.163682   
497   1.000000  0.401859  0.563219  0.578304 -0.126517 -0.058725  0.431806   
...        ...       ...       ...       ...       ...       ...       ...   
3772  1.000000  0.406366  0.268684  0.704863  0.358415  0.055376 -0.159812   
5191  1.000000  0.634337  0.594395  0.403271  0.004960  0.037880 -0.065399   
5226  1.000000  0.756413  0.732319  0.385807 -0.034821  0.073881 -0.058027   
5390  0.931410  0.626137  1.000000  0.375834 -0.223044  0.325943  0.121659   
860   1.000000  0.332793 -0.056006  0.358653  0.508890  0.185081 -0.205522   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_13  MFCCs_14  MFCCs_15  \
2004  0.146604  0.297338 -0.120902  ...  0.438532 -0.148451 -0.260375   
1194  0.261033 -0.168462 -0.281445  ... -0.108483  0.329984  0.274031   
5359 -0.064025  0.069728 -0.052552  ... -0.073050 -0.140994  0.162584   
1756 -0.014125  0.196527  0.018839  ...  0.404813 -0.075488 -0.320882   
497   0.066801 -0.321014  0.155569  ... -0.132244  0.190753  0.126674   
...        ...       ...       ...  ...       ...       ...       ...   
3772  0.065813  0.345068  0.175460  ...  0.378497 -0.014091 -0.352985   
5191  0.125276  0.106976 -0.223357  ... -0.228140 -0.274889  0.244973   
5226  0.183775  0.219419 -0.445112  ... -0.177708 -0.311815  0.182382   
5390 -0.092542  0.163365 -0.049382  ... -0.047610 -0.188463  0.102240   
860  -0.138280  0.204867  0.243058  ... -0.033355  0.255357  0.172043   

      MFCCs_16  MFCCs_17  MFCCs_18  MFCCs_19  MFCCs_20  MFCCs_21  MFCCs_22  
2004  0.178628  0.218024 -0.042083 -0.126511 -0.090279  0.096176  0.191697  
1194 -0.157579 -0.291979 -0.042932 -0.009021 -0.032679  0.023912  0.041407  
5359 -0.007981 -0.049428 -0.053166 -0.026728  0.029443  0.019032 -0.105221  
1756  0.024806  0.226971  0.034969 -0.106029 -0.147124  0.042627  0.250249  
497  -0.148049 -0.002891  0.157072 -0.015251 -0.055428  0.086927  0.086006  
...        ...       ...       ...       ...       ...       ...       ...  
3772  0.015183  0.220504  0.054246 -0.069108 -0.174583 -0.032202  0.188263  
5191  0.096130 -0.140923  0.003646  0.007163  0.068718  0.047640 -0.179215  
5226  0.030100 -0.182019  0.006296 -0.005868  0.029439  0.004783 -0.198014  
5390  0.150353 -0.054750 -0.054434  0.108092  0.156994 -0.051681 -0.085656  
860  -0.125121 -0.183615  0.003870  0.150144  0.040411 -0.103768 -0.066435  

[5036 rows x 22 columns]

In [4]:
train_y

Family      Genus                 Species
2004  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
1194    Dendrobatidae   Ameerega      Ameeregatrivittata
5359          Hylidae  Hypsiboas    HypsiboasCinerascens
1756  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
497   Leptodactylidae  Adenomera          AdenomeraAndre
...               ...        ...                     ...
3772  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
5191          Hylidae  Hypsiboas    HypsiboasCinerascens
5226          Hylidae  Hypsiboas    HypsiboasCinerascens
5390          Hylidae  Hypsiboas    HypsiboasCinerascens
860     Dendrobatidae   Ameerega      Ameeregatrivittata

[5036 rows x 3 columns]

In [5]:
test_x

MFCCs_ 1  MFCCs_ 2  MFCCs_ 3  MFCCs_ 4  MFCCs_ 5  MFCCs_ 6  MFCCs_ 7  \
3340       1.0  0.147506  0.014843  0.433466  0.169581  0.053135 -0.152020   
5108       1.0  0.674174  0.767520  0.392177  0.035669  0.181548 -0.013677   
6524       1.0 -0.007777  0.363276  0.413224  0.196296  0.222389  0.097128   
3649       1.0  0.429359  0.297881  0.609120  0.263991 -0.030414 -0.160778   
1617       1.0  0.190599  0.062234  0.617262  0.208825  0.027108 -0.214441   
...        ...       ...       ...       ...       ...       ...       ...   
1224       1.0  0.469993  0.342252  0.508362  0.184239  0.138370 -0.070231   
142        1.0  0.162558 -0.115829  0.045608  0.226107  0.325952  0.157953   
1490       1.0  0.160219  0.137078  0.599525  0.252240  0.077200 -0.132288   
2574       1.0  0.523548  0.387887  0.539775  0.087532  0.028962 -0.025403   
2653       1.0  0.368922  0.463250  0.569104  0.137126  0.074284 -0.015858   

      MFCCs_ 8  MFCCs_ 9  MFCCs_10  ...  MFCCs_13  MFCCs_14  MFCCs_15  \
3340 -0.102427  0.153061  0.128628  ...  0.341701  0.008463 -0.290680   
5108  0.187262  0.144679 -0.424524  ... -0.142909 -0.209932  0.285255   
6524 -0.155608 -0.035013  0.133689  ...  0.016052 -0.028317 -0.102525   
3649  0.076217  0.285909  0.052029  ...  0.374639  0.012647 -0.317314   
1617  0.009654  0.299053  0.039013  ...  0.402628 -0.110486 -0.306343   
...        ...       ...       ...  ...       ...       ...       ...   
1224  0.034221  0.226367 -0.001120  ...  0.256172 -0.229460 -0.247905   
142  -0.124050 -0.009148  0.235613  ... -0.318516  0.312135  0.401665   
1490 -0.001605  0.244840  0.081377  ...  0.443526 -0.090678 -0.323009   
2574  0.064718  0.215919 -0.005312  ...  0.342035 -0.160502 -0.156745   
2653  0.066424  0.277405  0.054899  ...  0.362272 -0.080797 -0.197053   

      MFCCs_16  MFCCs_17  MFCCs_18  MFCCs_19  MFCCs_20  MFCCs_21  MFCCs_22  
3340  0.016310  0.285756  0.166246 -0.061206 -0.191878  0.003221  0.255371  
5108  0.103722 -0.157943 -0.048913 -0.016129  0.129200  0.011245 -0.275909  
6524 -0.005594  0.125169  0.044882 -0.013309 -0.026086 -0.088825 -0.018968  
3649 -0.023621  0.230330  0.069015 -0.101196 -0.152351  0.039065  0.233823  
1617  0.082182  0.321598  0.128742 -0.152977 -0.223482  0.016131  0.250754  
...        ...       ...       ...       ...       ...       ...       ...  
1224  0.179489  0.186595 -0.080672 -0.104266 -0.077342  0.159712  0.239398  
142   0.120251 -0.120102 -0.226043 -0.161989  0.037390  0.185305  0.059824  
1490  0.082091  0.275059  0.046851 -0.108847 -0.190397 -0.011276  0.217408  
2574  0.175613  0.120708 -0.133983 -0.092651  0.011379  0.109798  0.100649  
2653  0.176598  0.200341 -0.040274 -0.113475 -0.055959  0.086084  0.169870  

[2159 rows x 22 columns]

In [6]:
test_y

Family      Genus                 Species
3340  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
5108          Hylidae  Hypsiboas    HypsiboasCinerascens
6524          Hylidae  Hypsiboas       HypsiboasCordobae
3649  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
1617  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
...               ...        ...                     ...
1224  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
142   Leptodactylidae  Adenomera          AdenomeraAndre
1490  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
2574  Leptodactylidae  Adenomera  AdenomeraHylaedactylus
2653  Leptodactylidae  Adenomera  AdenomeraHylaedactylus

[2159 rows x 3 columns]

(b) i. Research exact match and hamming score/ loss methods for evaluating multi- label classification and use them in evaluating the classifiers in this problem.

Exact match means the predicted set of labels should exactly match with the true set of labels, which is sometimes considered too strict. Hamming loss is the fraction of the wrong labels to the total number of labels. Both of these methods are to evaluation the accuracy of the predictions.

ii. Train a SVM for each of the labels, using Gaussian kernels and one versus all classifiers. Determine the weight of the SVM penalty and the width of the Gaussian Kernel using 10 fold cross validation. You are welcome to try to solve the problem with both standardized and raw attributes and report the results.

In [7]:
# Family 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10), 'kernel': ['rbf'],'gamma':np.arange(0.1,2,0.1)}]
svc = svm.SVC()
clf = GridSearchCV(svc, parameters,cv=10)
clf.fit(train_x,train_y['Family'])
clf.best_params_

{'C': 10.0, 'gamma': 1.9000000000000001, 'kernel': 'rbf'}

In [8]:
# Fit model 
svc = svm.SVC()
clf1 = make_pipeline(StandardScaler(), SVC(C=10.0,gamma=1.9000000000000001,kernel='rbf'))
clf1.fit(train_x, train_y['Family'])
predict_y_family=clf1.predict(test_x)
predict_y_family

array(['Leptodactylidae', 'Leptodactylidae', 'Hylidae', ...,
       'Leptodactylidae', 'Leptodactylidae', 'Leptodactylidae'],
      dtype=object)

In [10]:
# Genus
# Best params
parameters = [{'C': np.logspace(-3, 6, num = 10), 'kernel': ['rbf'],'gamma':np.arange(0.1,2,0.1)}]
svc = svm.SVC()
clf_genus = GridSearchCV(svc, parameters,cv=10)
clf_genus.fit(train_x,train_y['Genus'])
clf_genus.best_params_

{'C': 100.0, 'gamma': 1.5000000000000002, 'kernel': 'rbf'}

In [11]:
# Fit model 
svc = svm.SVC()
clf_genus1 = make_pipeline(StandardScaler(), SVC(C=100.0,gamma=1.5000000000000002,kernel='rbf'))
clf_genus1.fit(train_x, train_y['Genus'])
predict_y_genus=clf_genus1.predict(test_x)
predict_y_genus

array(['Adenomera', 'Adenomera', 'Hypsiboas', ..., 'Adenomera',
       'Adenomera', 'Adenomera'], dtype=object)

In [12]:
# Species
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10), 'kernel': ['rbf'],'gamma':np.arange(0.1,2,0.1)}]
svc = svm.SVC()
clf_spencies = GridSearchCV(svc, parameters,cv=10)
clf_spencies.fit(train_x,train_y['Species'])
clf_spencies.best_params_

{'C': 10.0, 'gamma': 1.8000000000000003, 'kernel': 'rbf'}

In [13]:
# Fit model 
svc = svm.SVC()
clf_species1 = make_pipeline(StandardScaler(), SVC(C=10.0,gamma=1.8000000000000003,kernel='rbf'))
clf_species1.fit(train_x, train_y['Species'])
predict_y_species=clf_species1.predict(test_x)
predict_y_species

array(['AdenomeraHylaedactylus', 'AdenomeraHylaedactylus',
       'HypsiboasCordobae', ..., 'AdenomeraHylaedactylus',
       'AdenomeraHylaedactylus', 'AdenomeraHylaedactylus'], dtype=object)

In [14]:
# EMR
def emr(y_true, y_pred):
    n = len(y_true)
    row_indicators = np.all(y_true == y_pred,axis = 1) # axis = 1 will check for equality along rows.
    exact_match_count = np.sum(row_indicators)
    return exact_match_count/n
y_true_rbf=np.array(test_y[['Family','Genus','Species']])
y_pred_rbf=[]                    
for i in range(len(predict_y_family)):
    y_pred_rbf.append([predict_y_family[i],predict_y_genus[i],predict_y_species[i]])
y_pred_rbf=np.array(y_pred_rbf,dtype=object)
emr_rbf=emr(y_true_rbf,y_pred_rbf)
emr_rbf

0.7114404817044928

In [15]:
# Exact Match Loss
eml_rbf=1-emr_rbf
eml_rbf

0.2885595182955072

In [16]:
# Hamming loss 
harmming_rbf=np.sum(np.not_equal(y_true_rbf, y_pred_rbf))/float(y_true_rbf.size)
harmming_rbf

0.21630384437239464

iii. Repeat 1(b)ii with L1-penalized SVMs. Remember to standardize4 the at- tributes. Determine the weight of the SVM penalty using 10 fold cross vali- dation.

In [17]:
# Standardize the attributes
train_x_stand = preprocessing.scale(train_x)
train_x_stand = pd.DataFrame(train_x_stand, columns = train_x.columns)
test_x_stand=preprocessing.scale(test_x)
test_x_stand = pd.DataFrame(test_x_stand, columns = test_x.columns)

In [18]:
# Family 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_family_stand = GridSearchCV(svc, parameters,cv=10)
clf_family_stand.fit(train_x_stand,train_y['Family'])
clf_family_stand.best_params_   #{'C': 1.0}

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

{'C': 1.0}

In [19]:
# Fit model 
svc = svm.SVC()
clf_family_stand1 = LinearSVC(penalty='l1', dual = False, C = 1.0)
clf_family_stand1.fit(train_x_stand,train_y['Family'])
predict_y_family_stand = clf_family_stand1.predict(test_x_stand)
predict_y_family_stand

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['Leptodactylidae', 'Hylidae', 'Hylidae', ..., 'Leptodactylidae',
       'Leptodactylidae', 'Leptodactylidae'], dtype=object)

In [20]:
# Genus 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_genus_stand = GridSearchCV(svc, parameters,cv=10)
clf_genus_stand.fit(train_x_stand,train_y['Genus'])
clf_genus_stand.best_params_  #{'C': 10.0}

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

{'C': 10.0}

In [21]:
# Fit model 
clf_genus_stand1 = LinearSVC(penalty='l1', dual = False, C = 10.0)
clf_genus_stand1.fit(train_x_stand,train_y['Genus'])
predict_y_genus_stand = clf_genus_stand1.predict(test_x_stand)
predict_y_genus_stand

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['Adenomera', 'Hypsiboas', 'Hypsiboas', ..., 'Adenomera',
       'Adenomera', 'Adenomera'], dtype=object)

In [22]:
# Species 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_species_stand = GridSearchCV(svc, parameters,cv=10)
clf_species_stand.fit(train_x_stand,train_y['Species'])
clf_species_stand.best_params_  #{'C': 10.0}

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

{'C': 10.0}

In [23]:
# Fit model 
clf_species_stand1 = LinearSVC(penalty='l1', dual = False, C = 10.0)
clf_species_stand1.fit(train_x_stand,train_y['Species'])
predict_y_species_stand = clf_species_stand1.predict(test_x_stand)
predict_y_species_stand

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['AdenomeraHylaedactylus', 'HypsiboasCinerascens',
       'HypsiboasCordobae', ..., 'AdenomeraHylaedactylus',
       'AdenomeraHylaedactylus', 'AdenomeraHylaedactylus'], dtype=object)

In [24]:
# EMR
y_true_linear=np.array(test_y[['Family','Genus','Species']])
y_pred_linear=[]                    
for i in range(len(predict_y_family_stand)):
    y_pred_linear.append([predict_y_family_stand[i],predict_y_genus_stand[i],predict_y_species_stand[i]])
y_pred_linear=np.array(y_pred_linear,dtype=object)
emr_linear=emr(y_true_linear,y_pred_linear)
emr_linear

0.9092172301991662

In [25]:
# Exact Match Loss
eml_linear=1-emr_linear
eml_linear

0.09078276980083377

In [26]:
# Hamming loss 
harmming_linear=np.sum(np.not_equal(y_true_linear, y_pred_linear))/float(y_true_linear.size)
harmming_linear

0.05866913694611703

iv. Repeat 1(b)iii by using SMOTE or any other method you know to remedy class imbalance. Report your conclusions about the classifiers you trained.

In [27]:
#SMOTE
sm = SMOTE(random_state=42)
train_x_family_res, train_y_family_res = sm.fit_resample(train_x, train_y['Family'])
train_x_genus_res, train_y_genus_res = sm.fit_resample(train_x, train_y['Genus'])
train_x_species_res, train_y_species_res = sm.fit_resample(train_x, train_y['Species'])

In [ ]:
# Family 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_family_sm = GridSearchCV(svc, parameters,cv=10)
clf_family_sm.fit(train_x_family_res,train_y_family_res)
clf_family_sm.best_params_   #{'C': 100000.0}

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

In [29]:
# Fit model 
clf_family_sm1 = LinearSVC(penalty='l1', dual = False, C = 100000.0)
clf_family_sm1.fit(train_x_family_res,train_y_family_res)
predict_y_family_sm = clf_family_sm1.predict(test_x)
predict_y_family_sm

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['Leptodactylidae', 'Hylidae', 'Hylidae', ..., 'Leptodactylidae',
       'Leptodactylidae', 'Leptodactylidae'], dtype=object)

In [30]:
# Genus 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_genus_sm = GridSearchCV(svc, parameters,cv=10)
clf_genus_sm.fit(train_x_genus_res,train_y_genus_res)
clf_genus_sm.best_params_   

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

{'C': 100.0}

In [31]:
# Fit model 
clf_genus_sm1 = LinearSVC(penalty='l1', dual = False, C = 100000)
clf_genus_sm1.fit(train_x_genus_res,train_y_genus_res)
predict_y_genus_sm = clf_genus_sm1.predict(test_x)
predict_y_genus_sm

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['Adenomera', 'Hypsiboas', 'Hypsiboas', ..., 'Adenomera',
       'Adenomera', 'Adenomera'], dtype=object)

In [32]:
# Species 
# Best params 
parameters = [{'C': np.logspace(-3, 6, num = 10)}]
svc = svm.LinearSVC(penalty='l1',dual=False)
clf_species_sm = GridSearchCV(svc, parameters,cv=10)
clf_species_sm.fit(train_x_species_res,train_y_species_res)
clf_species_sm.best_params_  #{'C': 1000.0}

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number

{'C': 1000.0}

In [33]:
# Fit model 
clf_species_sm1 = LinearSVC(penalty='l1', dual = False, C = 1000.0)
clf_species_sm1.fit(train_x_species_res,train_y_species_res)
predict_y_species_sm = clf_species_sm1.predict(test_x)
predict_y_species_sm

/usr/local/lib/python3.7/dist-packages/sklearn/svm/_base.py:1208: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  ConvergenceWarning,


array(['AdenomeraHylaedactylus', 'HypsiboasCinerascens',
       'HypsiboasCordobae', ..., 'AdenomeraHylaedactylus',
       'AdenomeraHylaedactylus', 'AdenomeraHylaedactylus'], dtype=object)

In [34]:
# EMR
y_true_sm=np.array(test_y[['Family','Genus','Species']])
y_pred_sm=[]                    
for i in range(len(predict_y_family_sm)):
    y_pred_sm.append([predict_y_family_sm[i],predict_y_genus_sm[i],predict_y_species_sm[i]])
y_pred_sm=np.array(y_pred_sm,dtype=object)
emr_sm=emr(y_true_sm,y_pred_sm)
emr_sm

0.857341361741547

In [35]:
# Exact Match Loss
eml_sm=1-emr_sm
eml_sm

0.14265863825845304

In [36]:
# Hamming loss 
harmming_sm=np.sum(np.not_equal(y_true_sm, y_pred_sm))/float(y_true_sm.size)
harmming_sm

0.0745715609078277

As we can see the hamming loss= 0.0745715609078277 is way less than the Eact Match loss= 0.14265863825845304, which indicated this this is a better model.

## 2. K-Means Clustering on a Multi-Class and Multi-Label Data Set

(a) Use k-means clustering on the whole Anuran Calls (MFCCs) Data Set (do not split the data into train and test, as we are not performing supervised learning in this exercise). Choose k ∈ {1, 2, . . . , 50} automatically based on one of the methods provided in the slides (CH or Gap Statistics or scree plots or Silhouettes) or any other method you know.

In [42]:
x=data.drop(['Family','Genus','Species','RecordID'],axis=1)
y=data[['Family','Genus','Species']]

In [45]:
for i in range(1,51): # Monte-Carlo Simulation
    d=[]
    for k in range(2,51): 
        clusterer = KMeans(n_clusters=k)
        cluster_labels = clusterer.fit_predict(x)
        # Silhouettes scores
        silhouette=silhouette_score(x, cluster_labels,metric='euclidean')
        d.append(silhouette)
    print('max silhouette', max(d))
    print('k =',d.index(max(d))+2) # Because our k starts at 2 and the first index is 0, we add 2 here to get the real k
    

max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.37863353074850936
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.38523395202479643
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.37863353074850936
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3788646212055977
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3788974185180753
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.37866647707550904
k = 4
max silhouette 0.3746938723900754
k = 4
max silhouette 0.37863353074850936
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.3787509343305295
k = 4
max silhouette 0.374709713992453
k = 4
max silhouette 0.37863353074850936
k = 4
max silhouette 0.3787509343305295
k

From the result above, we can see that the optimal cluster number is 4.

(b) In each cluster, determine which family is the majority by reading the true labels. Repeat for genus and species.


In [84]:
#from scipy.spatial.distance import hamming
hamming_loss_score=[]
hamming_score=[]
hamming_distance=[]
for t in range(1,51):
    clusterer = KMeans(n_clusters=4)
    cluster_labels = clusterer.fit_predict(x) 
    data['predict'] = cluster_labels.tolist()
    predict = data['predict'].to_list()
    for k in range(0,4):
        labels=data[data['predict']==k]
        family_label=labels['Family'].value_counts().index[0]
        genus_label=labels['Genus'].value_counts().index[0]
        species_label=labels['Species'].value_counts().index[0]
        result=[family_label,genus_label,species_label]
        for i in predict:
            if i ==k:
                predict=result
            else:
                i
    predict=np.array(predict, dtype = 'object')
    print(predict)
    hamming_loss=np.sum(np.not_equal(y, predict))/float(y.size)
    hamming_loss_score.append(hamming_loss) 
    hamming_score.append(1-hamming_loss)
    hamming_distance.append(hamming_loss*3)

['Dendrobatidae' 'Ameerega' 'Ameeregatrivittata']
['Leptodactylidae' 'Adenomera' 'AdenomeraHylaedactylus']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Dendrobatidae' 'Ameerega' 'Ameeregatrivittata']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Hylidae' 'Hypsiboas' 'HypsiboasCinerascens']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Leptodactylidae' 'Adenomera' 'AdenomeraHylaedactylus']
['Hylidae' 'Hypsiboas' 'HypsiboasCinerascens']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Dendrobatidae' 'Ameerega' 'Ameeregatrivittata']
['Dendrobatidae' 'Ameerega' 'Ameeregatrivittata']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Leptodactylidae' 'Adenomera' 'AdenomeraHylaedactylus']
['Hylidae' 'Hypsiboas' 'HypsiboasCordobae']
['Leptodactylidae' 'Adenomera' 'AdenomeraHylaedactylus']
['Dendrobatidae' 'Ameerega' 'Ameeregatrivittata']
['Hylidae' 'Hypsiboas' 'HypsiboasCinerascens']
['Hylidae' 'Hypsiboas' 'Hypsi

(c) Now for each cluster you have a majority label triplet (family, genus, species). Calculate the average Hamming distance, Hamming score, and Hamming loss between the true labels and the labels assigned by clusters.


In [85]:
avg_hamming_score=np.mean(hamming_score)
avg_hamming_score

0.7601550459423984

In [86]:
std_hamming_score=np.std(hamming_score)
std_hamming_score

0.0639250545072141

In [87]:
hamming_score

[Family     0.691777
 Genus      0.691777
 Species    0.691777
 dtype: float64, Family     0.871438
 Genus      0.858930
 Species    0.827797
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.691777
 Genus      0.691777
 Species    0.691777
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.688534
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.871438
 Genus      0.858930
 Species    0.827797
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.688534
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.766968
 Genus      0.740468
 Species    0.718601
 dtype: float64, Family     0.691777
 Genus      0.691777
 Species  

In [88]:
avg_hamming_distance=np.mean(hamming_distance)
avg_hamming_distance

0.7195348621728052

In [89]:
std_hamming_distance=np.std(hamming_distance)
std_hamming_distance

0.1917751635216424

In [90]:
hamming_distance

[Family     0.92467
 Genus      0.92467
 Species    0.92467
 dtype: float64, Family     0.385685
 Genus      0.423211
 Species    0.516609
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.92467
 Genus      0.92467
 Species    0.92467
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.934399
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.385685
 Genus      0.423211
 Species    0.516609
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.934399
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.699097
 Genus      0.778596
 Species    0.844197
 dtype: float64, Family     0.92467
 Genus      0.92467
 Species    0.9246

In [91]:
avg_hamming_loss=np.mean(hamming_loss_score)
avg_hamming_loss

0.23984495405760176

In [92]:
std_hamming_loss=np.std(hamming_loss_score)
std_hamming_loss

0.06392505450721413

In [93]:
hamming_loss_score

[Family     0.308223
 Genus      0.308223
 Species    0.308223
 dtype: float64, Family     0.128562
 Genus      0.141070
 Species    0.172203
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.308223
 Genus      0.308223
 Species    0.308223
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.311466
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.128562
 Genus      0.141070
 Species    0.172203
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.311466
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.233032
 Genus      0.259532
 Species    0.281399
 dtype: float64, Family     0.308223
 Genus      0.308223
 Species  

## Reference

1.(b)ii https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

1.(b)ii https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html

1.(b)ii https://medium.datadriveninvestor.com/a-survey-of-evaluation-metrics-for-multilabel-classification-bb16e8cd41cd


1.(b)ii https://stackoverflow.com/questions/38697982/python-scikit-learn-multi-class-multi-label-performance-metrics


1.(b)iii https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html


1.(b)iii https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html


1.(b)iv https://imbalanced-learn.org/stable/references/generated/imblearn.over_sampling.SMOTE.html


1.(b)iv https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html


3.(a) https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html
